### DISCUSSION TITLE
How bad if I don't offer a surface when creating a device?

### DISCUSSION THREAD

**Discussion by psionic12** [gfx-rs/wgpu#1849] [Q&A] [answered]
Hi,
For some reason, I want to create a device before any window exists. In order to do this, I have to write like this:
```
        let adapter = instance.request_adapter(
            &wgpu::RequestAdapterOptions {
                power_preference: wgpu::PowerPreference::HighPerformance,
                compatible_surface: None,
            },
```

It works just fine on my machine, windows created later do rendering images, though I didn't test the code thoroughly. I wondered how bad is that to set `compatible_surface: None`?

And if it is bad, is it a good idea to create a window that doing nothing (even not showing), just for creating a device?

**Comment by kvark** [comment, marked as answer]
The surface provided there isn't required by the adapter/device. The relationship is separate: if you have a surface that you want to present on, you'd specify it here in order to guarantee that you *can* present on this surface. Without it provided at adapter creation, there is easily a situation where the surface can't be presented on some of the machines.

**Reply by psionic12** [reply]
Sound harmless to me since I'm creating a queue just for transferring, right? 
I did this in vulkan, except that I will explicitly finding a transfer queue family, while I didn't find the related API in wgpu.

**Reply by cwfitzgerald** [reply]
Correct, if you're not rendering to a surface, you don't need to worry about that.

As for multiple queue, we don't currently support it, so you can use the primary queue for everything.

**Reply by psionic12** [reply]
That would be nice if multiple queues feature is support. It would improve the performance of wgpu on some high-end graphic cards.

**Reply by cwfitzgerald** [reply]
It's definitely on our radar (there's an issue open for it) but it's going to require a massive amount of work to get right.
